ext/spl/spl_fixedarray.c

```diff
@@ -208,6 +263,19 @@ static HashTable* spl_fixedarray_object_get_properties(zend_object *obj)
 		zend_long j = zend_hash_num_elements(ht);
 
 		if (GC_REFCOUNT(ht) > 1) {
+			/*
+			 * Usually, the reference count of the hash table is 1,
+			 * except during cyclic reference cycles.
+			 *
+			 * Attempt to maintain the DEBUG invariant that a hash table isn't modified during iteration.
+			 *
+			 * See https://github.com/php/php-src/issues/8079 and ext/spl/tests/fixedarray_022.phpt
+			 * Also see https://github.com/php/php-src/issues/8044 for alternate considered approaches.
+			 */
+			if (!spl_fixedarray_object_needs_rebuild(intern, ht, j)) {
+				/* Return the same hash table so that recursion cycle detection works in internal functions. */
+				return ht;
+			}
```

Comment:
Interesting approach, but I afraid it's too expensive. I thought about setting some flag on SplFixedArray modification. 

BTW this leads me to another idea - we may try to return the same ``ht`` if it's already protected from recursion.

```c
    if (GC_IS_RECURSIVE(ht)) {
        return ht;
    }
```
Ideally, we should switch from ``get_properties`` to ``get_properties_for`` and do this for everything except ``ZEND_PROP_PURPOSE_ARRAY_CAST``. Thanks for persistence. I hope, we will fix this soon.

Now, I also see a mesh in ``ext/standard/var.c`` that uses different protection mechanisms in different places. Did you propose a PR that tried to change this? I suppose, ``Z_IS_RECURSIVE_P/GC_IS_RECURSIVE`` were used on purpose, but I can't remember anything. @nikic do you have any thoughts?

Reply:
> Now, I also see a mesh in ext/standard/var.c that uses different protection mechanisms in different places. Did you propose a PR that tried to change this? 

Yes, I'd proposed https://github.com/php/php-src/pull/8046 `Make var_export/debug_zval_dump first check for infinite recursion on the object` to try to change this to always check for infinite recursion on the **object** before get_properties/get_properties_for. Could you review that?

> I suppose, `Z_IS_RECURSIVE_P`/`GC_IS_RECURSIVE` were used on purpose, but I can't remember anything.

I also don't know, brainstorming ideas of why:

1. Trying to get mixes of json_encode calling var_export to work instead of failing with `RECURSION`? - seems obscure and not likely. There are alternate approaches even if that really was the approach, e.g. new `GC_FLAGS` bits specifically for var_export/debug dumpers
2. Trying to detect sharing the same hash table - I can't think of anything doing that

Reply:
> BTW this leads me to another idea - we may try to return the same ht if it's already protected from recursion.
> 
> ```
>     if (GC_IS_RECURSIVE(ht)) {
>         return ht;
>     }
> ```

It's protected from recursion when debugging - var_export/var_dump would show the wrong results if the side effects of `__debugInfo`, `jsonSerialize`, etc. of another method modified the underlying table during a json_encode/var_dump call, but it wouldn't crash or misbehave. 

That's probably work in combination with adding an implementation of `get_properties_for` for other `ZEND_PROP_PURPOSE_*`, **though with all the problems this has had so far, I'd rather introduce that in 8.2 rather than in a patch release.** (especially with SplFixedArray being subclassable and subclasses being able to override `JsonSerializable`, `__sleep`, `__debugInfo`, etc.

(because json_encode, var_export, debug_zval_dump wouldn't use the hash table, they'd just report recursion).

Reply:
>Interesting approach, but I afraid it's too expensive. I thought about setting some flag on SplFixedArray modification.

I pushed a new commit doing that instead